### PR TITLE
mds: add root_squash mode in MDS auth caps

### PR DESCRIPTION
**Pull request by ajarr**
Implement a root_squash mode in MDS auth caps to deny operations for
    clients with uid=0 or gid=0 that need write access. It's mainly to
    prevent operations such as accidental `sudo rm -rf /path`.
    
    The root squash mode can be enforced in one of the following ways in
    the MDS caps,
      'allow rw root_squash'
      (across file systems)
              or
      'allow rw fsname=a root_squash'
      (on a file system)
              or
      'allow rw fsname=a path=/vol/group/subvol00 root_squash'
      (on a file system path)
    
    Fixes: https://tracker.ceph.com/issues/42451
    Signed-off-by: Ramana Raja <rraja@redhat.com>

### CODE REVIEW

**Comment by ajarr** [comment]
This PR is based on the [multiple FS auth ID PR](https://github.com/ceph/ceph/pull/32581) that's not been merged. Following is the only commit
https://github.com/ceph/ceph/pull/36457/commits/1f5d61a040f18930d4e9c358123e116c540b98ce
that enables root_squash, the other commits are part of the multiple FS auth ID PR.

File: src/mds/MDSAuthCaps.cc [hunk 1]
```diff
@@ -70,11 +71,16 @@ struct MDSCapParser : qi::grammar<Iterator, MDSAuthCaps()>
     uintlist %= (uint_ % lit(','));
     gidlist %= -(spaces >> lit("gids") >> lit('=') >> uintlist);
     fs_name %= -(spaces >> lit("fsname") >> lit('=') >> fs_name_str);
+    root_squash %= -(spaces >> lit("root_squash") >> lit("=") >> bool_);
```
Comment (batrick) [inline src/mds/MDSAuthCaps.cc:74]: I don't think there's any need for this bool. Can't it just be a flag?

File: src/mds/MDSAuthCaps.cc [hunk 2]
```diff
@@ -231,6 +238,11 @@ bool MDSAuthCaps::is_capable(std::string_view inode_path,
 
     if (grant.match.match(inode_path, caller_uid, caller_gid, caller_gid_list) &&
 	grant.spec.allows(mask & (MAY_READ|MAY_EXECUTE), mask & MAY_WRITE)) {
+      if (grant.match.root_squash && ((caller_uid == 0) || (caller_gid == 0))) {
+	  if (mask & MAY_WRITE) {
```
Comment (batrick) [inline src/mds/MDSAuthCaps.cc:242]: merge `if` statements

File: src/mds/MDSAuthCaps.cc [hunk 3]
```diff
@@ -361,12 +373,18 @@ bool MDSAuthCaps::allow_all() const
...
+    out << "fs_name=" << match.fs_name << " ";
+  }
   if (match.path.length()) {
     out << "path=\"/" << match.path << "\"";
     if (match.uid != MDSCapMatch::MDS_AUTH_UID_ANY) {
       out << " ";
     }
   }
+  if (match.root_squash) {
+    out << "root_squash=" << match.root_squash << " ";
```
Comment (batrick) [inline src/mds/MDSAuthCaps.cc:386]:
```suggestion
    out << "root_squash ";
```

File: src/mds/MDSAuthCaps.h [hunk 4]
```diff
@@ -145,6 +151,7 @@ struct MDSCapMatch {
   std::vector<gid_t> gids;  // Use these GIDs
   std::string path;  // Require path to be child of this (may be "" or "/" for any)
   std::string fs_name;
+  bool root_squash;
```
Comment (batrick) [inline src/mds/MDSAuthCaps.h:154]:
```suggestion
  bool root_squash = false;
```

**Comment by ajarr** [comment]
This PR is based on the [multiple FS auth ID PR](https://github.com/ceph/ceph/pull/32581) that's not been merged. Following is the only commit
https://github.com/ceph/ceph/pull/36457/commits/1f5d61a040f18930d4e9c358123e116c540b98ce
that enables root_squash, the other commits are part of the multiple FS auth ID PR.

**Review by batrick — DISMISSED** [review]
This looks very good. We just need tests for `fs authorize` with `root_squash`. Please add them to `test_admin.py`. You already have tests for checking the capability so I think you can skip validating the caps work with a mount.

**Comment by ajarr** [comment]
> This looks very good. We just need tests for `fs authorize` with `root_squash`. Please add them to `test_admin.py`. You already have tests for checking the capability so I think you can skip validating the caps work with a mount.

Done.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test make check

**Comment by ajarr** [comment]
jenkins test make check

**Comment by ajarr** [comment]
@batrick do you want me to test this in teuthology?

**Comment by batrick** [comment]
@ajarr I encourage testing your own PR for the specific tests you're adding. I'll do broader QA testing in the next few days.

**Comment by ajarr** [comment]
testing specific tests that were modified by this PR
https://pulpito.ceph.com/rraja-2020-10-06_10:29:44-fs-wip-ajarr-root-squash-distro-basic-smithi/

export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-ajarr-root-squash
teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -p 50 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH --filter admin,libcephfs_interface_tests --filter-all objectstore-ec/bluestore-bitmap

"--filter admin,libcephfs_interface_tests --filter-all objectstore-ec/bluestore-bitmap" are the filters used.

**Comment by ajarr** [comment]
> tests for changes in TestMDSAuthCaps.cc
> https://pulpito.ceph.com/rraja-2020-10-06_10:29:44-fs-wip-ajarr-root-squash-distro-basic-smithi/
> 
> tests for changes in test_admin
> https://pulpito.ceph.com/rraja-2020-10-06_10:34:06-fs-wip-ajarr-root-squash-distro-basic-smithi/

Tests passed.

**Comment by batrick** [comment]
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20201007.214100

failures unrelated.

**Pull request merged**